File: +json/readers/mloads.jl
"""
    MLoads

Read MATLAB `json.mdumps` format-2 payloads in Julia. Reference implementation
of the format in `+json/FORMAT.md`. Read-only: there is no Julia writer.

Deliberately **parser-agnostic** — it takes an already-parsed JSON value, so it
adds no dependency of its own and works with whichever JSON package a project
already has:

```julia
include("mloads.jl")
using .MLoads, JSON            # or JSON3

obj = mloads(JSON.parse(row_text))            # JSON.jl
obj = mloads(JSON3.read(row_text, Any))       # JSON3.jl
```

Type mapping
------------
| MATLAB                    | Julia                                          |
|:--------------------------|:-----------------------------------------------|
| scalar struct             | `Dict{String,Any}`                             |
| struct array              | `Vector{Dict{String,Any}}`, column-major       |
| cell array                | `Vector{Any}`, column-major                    |
| numeric / logical array   | `Array{T,N}` with MATLAB's shape               |
| numeric / logical scalar  | `Float64` / `Int64` / `Bool` / ...             |
| char (row vector, empty)  | `String`                                       |
| char (2-D, >1 row)        | `Vector{String}`, one per row                  |
| string array              | `Vector{String}` (`String` if scalar)          |

Multidimensional cell and struct arrays come back as column-major vectors;
`mloads(x; with_meta=true)` returns `(value, meta)` with every node's MATLAB
class and dims if you need the shape.

Leaves are stored flattened column-major, which is Julia's own memory order, so
`reshape` is correct as written and needs no permuting.
"""
module MLoads

export mloads, MLoadsError

struct MLoadsError <: Exception
    msg::String
end
Base.showerror(io::IO, e::MLoadsError) = print(io, "MLoadsError: ", e.msg)

const TYPES = Dict{String,DataType}(
    "double"  => Float64,
    "single"  => Float32,
    "int8"    => Int8,
    "uint8"   => UInt8,
    "int16"   => Int16,
    "uint16"  => UInt16,
    "int32"   => Int32,
    "uint32"  => UInt32,
    "int64"   => Int64,
    "uint64"  => UInt64,
    "logical" => Bool,
)

"""
    mloads(parsed; with_meta=false)

Rebuild a MATLAB value from a parsed `json.mdumps` payload. Pass the result of
`JSON.parse` / `JSON3.read(..., Any)`, not the raw text.
"""
function mloads(parsed; with_meta::Bool = false)
    obj = _asdict(parsed)
    if obj === nothing || !haskey(obj, "vals") || !haskey(obj, "info")
        throw(MLoadsError("not a json.mdumps payload (no vals/info)"))
    end

    if !haskey(obj, "fmt")
        @warn """format-1 (legacy) payload: returning "vals" as plain JSON \
                 without restoring MATLAB types or shapes. Re-save it with the \
                 current json.mdumps for a faithful decode."""
        return with_meta ? (obj["vals"], Dict()) : obj["vals"]
    end
    Int(obj["fmt"]) == 2 || throw(MLoadsError("unsupported format version $(obj["fmt"])"))

    info = obj["info"]
    info isa AbstractVector || throw(MLoadsError("info must be a list"))

    value, idx = _build(obj["vals"], info, 1)
    idx == length(info) + 1 ||
        @warn "consumed $(idx - 1) of $(length(info)) info entries; payload may be malformed"

    if with_meta
        meta = Dict(Tuple(_aslist(get(e, "p", []))) =>
                    (String(e["t"]), Tuple(Int.(_aslist(e["d"])))) for e in _asdict.(info))
        return value, meta
    end
    return value
end

# ---------------------------------------------------------------------------
# JSON3 hands back its own object type rather than a Dict, so normalise once
# here instead of special-casing the parser everywhere below.
_asdict(x::AbstractDict) = x
function _asdict(x)
    if !(x isa AbstractVector) && !(x isa AbstractString) && !(x isa Number) &&
       x !== nothing && applicable(keys, x) && applicable(getindex, x, :a)
        return Dict{String,Any}(String(k) => x[k] for k in keys(x))
    end
    return nothing
end

"""A JSON scalar stands in for a one-element array; normalise to a vector."""
_aslist(x::AbstractVector) = collect(x)
_aslist(::Nothing) = Any[]
_aslist(x) = Any[x]

"""The `n` raw child blocks of a container."""
function _elements(raw, n::Int)
    n == 0 && return Any[]
    if raw isa AbstractVector
        length(raw) == n && return collect(raw)
        n == 1 && return Any[raw]
        throw(MLoadsError("expected $n container elements, found $(length(raw))"))
    end
    n == 1 && return Any[raw]
    raw === nothing && return Any[nothing for _ in 1:n]
    throw(MLoadsError("expected $n container elements, found a $(typeof(raw))"))
end

function _build(raw, info, i::Int)
    i <= length(info) || throw(MLoadsError("ran out of info entries"))
    e = _asdict(info[i])
    e === nothing && throw(MLoadsError("info entry $i is not an object"))
    i += 1

    dims = Int.(_aslist(e["d"]))
    n = prod(dims)
    t = String(e["t"])
    get(e, "as", nothing) == "struct" && (t = "struct")

    if t == "cell"
        kids = _elements(raw, n)
        out = Vector{Any}(undef, n)
        for k in 1:n
            out[k], i = _build(kids[k], info, i)
        end
        return out, i

    elseif t == "struct"
        fields = String.(_aslist(get(e, "f", [])))
        if n == 1
            src = _asdict(raw)
            out = Dict{String,Any}()
            for name in fields
                v, i = _build(src === nothing ? nothing : get(src, name, nothing), info, i)
                out[name] = v
            end
            return out, i
        end
        kids = _elements(raw, n)
        out = Vector{Dict{String,Any}}(undef, max(n, 0))
        for k in 1:n
            src = _asdict(kids[k])
            elem = Dict{String,Any}()
            for name in fields
                v, i = _build(src === nothing ? nothing : get(src, name, nothing), info, i)
                elem[name] = v
            end
            out[k] = elem
        end
        return out, i

    elseif t == "char"
        s = raw isa AbstractString ? String(raw) : ""
        if length(dims) == 2 && dims[1] > 1
            rows, cols = dims[1], dims[2]
            ch = collect(s)
            length(ch) < rows * cols && append!(ch, fill(' ', rows * cols - length(ch)))
            # stored column-major: element (r, c) sits at r + rows * (c - 1)
            return [String([ch[r + rows * (c - 1)] for c in 1:cols]) for r in 1:rows], i
        end
        return s, i

    elseif t == "string"
        items = [x isa AbstractString ? String(x) : "" for x in _aslist(raw)]
        while length(items) < n
            push!(items, "")
        end
        items = items[1:n]
        return n == 1 ? items[1] : items, i

    elseif haskey(TYPES, t)
        T = TYPES[t]
        flat = _numeric_leaf(raw, e, n, T)
        if n == 1 && all(==(1), dims)
            return flat[1], i
        end
        return reshape(flat, dims...), i
    end

    throw(MLoadsError("unknown MATLAB class \"$t\" in info"))
end

function _numeric_leaf(raw, e, n::Int, ::Type{T}) where {T}
    if haskey(e, "s") && !isempty(_aslist(e["s"]))
        # exact decimal strings for 64-bit ints beyond 2^53
        return T[parse(T, String(x)) for x in _aslist(e["s"])][1:n]
    end

    vals = _aslist(raw)
    flat = Vector{T}(undef, n)
    filler = T <: AbstractFloat ? T(NaN) : zero(T)
    for k in 1:n
        if k <= length(vals) && vals[k] !== nothing
            flat[k] = T <: Bool ? T(vals[k] != 0) : T(vals[k])
        else
            flat[k] = filler
        end
    end

    nf = _asdict(get(e, "nf", nothing))
    if nf !== nothing && T <: AbstractFloat
        idxs = _aslist(get(nf, "i", []))
        kinds = _aslist(get(nf, "k", []))
        for (pos, kind) in zip(idxs, kinds)
            j = Int(pos)
            1 <= j <= n || continue
            k = String(kind)
            flat[j] = k == "NaN" ? T(NaN) : k == "Inf" ? T(Inf) : k == "-Inf" ? T(-Inf) : flat[j]
        end
    end
    return flat
end

end # module

File: +json/readers/mloads.R
# mloads.R -- read MATLAB json.mdumps format-2 payloads in R.
#
# Reference implementation of the format in +json/FORMAT.md.
# Read-only: there is no R writer.
#
#   source("mloads.R")
#   obj <- mloads(row_text_from_mariadb)          # needs jsonlite
#   obj <- mloads_parsed(already_parsed)          # no dependency at all
#
# Type mapping
# ------------
#   scalar struct               named list
#   struct array                unnamed list of named lists, column-major
#   cell array                  unnamed list, column-major
#   numeric / logical array     array/matrix/vector with MATLAB's shape
#   numeric / logical scalar    length-1 vector
#   char (row vector, empty)    character(1)
#   char (2-D, >1 row)          character vector, one element per row
#   string array                character vector
#
# Multidimensional cell and struct arrays come back as column-major lists;
# mloads(x, with_meta = TRUE) also returns every node's MATLAB class and dims.
#
# Leaves are stored flattened column-major, which is R's own array order, so
# array() and dim() are correct as written and need no transposing.
#
# IMPORTANT: parse with simplifyVector = FALSE. jsonlite's default
# simplification collapses the nested arrays that carry cell structure, and
# turns nulls into different shapes depending on the data.

mloads <- function(text, with_meta = FALSE) {
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    stop("mloads(): package 'jsonlite' is required to parse text; ",
         "use mloads_parsed() with your own parser instead.", call. = FALSE)
  }
  parsed <- jsonlite::fromJSON(text, simplifyVector = FALSE)
  mloads_parsed(parsed, with_meta = with_meta)
}

mloads_parsed <- function(obj, with_meta = FALSE) {
  if (!is.list(obj) || !all(c("vals", "info") %in% names(obj))) {
    stop("not a json.mdumps payload (no vals/info)", call. = FALSE)
  }

  if (!("fmt" %in% names(obj))) {
    warning("format-1 (legacy) payload: returning 'vals' as plain JSON without ",
            "restoring MATLAB types or shapes. Re-save it with the current ",
            "json.mdumps for a faithful decode.", call. = FALSE)
    return(if (with_meta) list(value = obj[["vals"]], meta = list()) else obj[["vals"]])
  }
  if (as.integer(obj[["fmt"]]) != 2L) {
    stop("unsupported format version ", obj[["fmt"]], call. = FALSE)
  }

  info <- obj[["info"]]
  if (!is.list(info)) stop("info must be a list", call. = FALSE)

  res <- .build(obj[["vals"]], info, 1L)
  if (res$i != length(info) + 1L) {
    warning("consumed ", res$i - 1L, " of ", length(info),
            " info entries; payload may be malformed", call. = FALSE)
  }

  if (!with_meta) return(res$value)

  meta <- lapply(info, function(e) {
    list(path = .aslist(e[["p"]]),
         class = e[["t"]],
         dims = as.integer(unlist(.aslist(e[["d"]]))))
  })
  list(value = res$value, meta = meta)
}

# --- internals -------------------------------------------------------------

.MATLAB_NUMERIC <- c("double", "single", "int8", "uint8", "int16", "uint16",
                     "int32", "uint32", "int64", "uint64", "logical")

# A JSON scalar stands in for a one-element array; normalise to a list.
.aslist <- function(x) {
  if (is.null(x)) return(list())
  if (is.list(x)) return(x)
  as.list(x)
}

# The n raw child blocks of a container.
.elements <- function(raw, n) {
  if (n == 0L) return(list())
  if (is.list(raw) && is.null(names(raw))) {
    if (length(raw) == n) return(raw)
    if (n == 1L) return(list(raw))
    stop("expected ", n, " container elements, found ", length(raw), call. = FALSE)
  }
  if (n == 1L) return(list(raw))
  if (is.null(raw)) return(rep(list(NULL), n))
  stop("expected ", n, " container elements, found a ", class(raw)[1], call. = FALSE)
}

.build <- function(raw, info, i) {
  if (i > length(info)) stop("ran out of info entries", call. = FALSE)
  e <- info[[i]]
  i <- i + 1L

  dims <- as.integer(unlist(.aslist(e[["d"]])))
  n <- prod(dims)
  t <- e[["t"]]
  if (identical(e[["as"]], "struct")) t <- "struct"

  if (identical(t, "cell")) {
    kids <- .elements(raw, n)
    out <- vector("list", n)
    for (k in seq_len(n)) {
      r <- .build(kids[[k]], info, i)
      out[[k]] <- r$value
      i <- r$i
    }
    return(list(value = out, i = i))
  }

  if (identical(t, "struct")) {
    fields <- unlist(.aslist(e[["f"]]))
    if (n == 1L) {
      src <- if (is.list(raw)) raw else list()
      out <- list()
      for (nm in fields) {
        r <- .build(src[[nm]], info, i)
        out[[nm]] <- r$value
        i <- r$i
      }
      return(list(value = out, i = i))
    }
    kids <- .elements(raw, n)
    out <- vector("list", n)
    for (k in seq_len(n)) {
      src <- if (is.list(kids[[k]])) kids[[k]] else list()
      elem <- list()
      for (nm in fields) {
        r <- .build(src[[nm]], info, i)
        elem[[nm]] <- r$value
        i <- r$i
      }
      out[[k]] <- elem
    }
    return(list(value = out, i = i))
  }

  if (identical(t, "char")) {
    s <- if (is.character(raw) && length(raw) == 1L) raw else ""
    if (length(dims) == 2L && dims[1] > 1L) {
      rows <- dims[1]; cols <- dims[2]
      ch <- strsplit(s, "", fixed = TRUE)[[1]]
      if (length(ch) < rows * cols) ch <- c(ch, rep(" ", rows * cols - length(ch)))
      # stored column-major: element (r, c) sits at r + rows * (c - 1)
      rowstr <- vapply(seq_len(rows), function(r) {
        paste(ch[r + rows * (seq_len(cols) - 1L)], collapse = "")
      }, character(1))
      return(list(value = rowstr, i = i))
    }
    return(list(value = s, i = i))
  }

  if (identical(t, "string")) {
    items <- vapply(.aslist(raw), function(x) {
      if (is.character(x) && length(x) == 1L) x else ""
    }, character(1))
    if (length(items) < n) items <- c(items, rep("", n - length(items)))
    return(list(value = items[seq_len(n)], i = i))
  }

  if (!(t %in% .MATLAB_NUMERIC)) {
    stop('unknown MATLAB class "', t, '" in info', call. = FALSE)
  }

  flat <- .numeric_leaf(raw, e, n, t)
  if (n == 1L && all(dims == 1L)) return(list(value = flat, i = i))
  # R arrays are column-major, matching the stored order, so no permuting.
  return(list(value = array(flat, dim = dims), i = i))
}

.numeric_leaf <- function(raw, e, n, t) {
  logical_leaf <- identical(t, "logical")

  if (!is.null(e[["s"]]) && length(.aslist(e[["s"]])) > 0L) {
    # Exact decimal strings for 64-bit ints beyond 2^53. R has no native int64,
    # so these come back as doubles and lose precision above 2^53. Use the
    # bit64 package on the strings if you need them exactly.
    strs <- vapply(.aslist(e[["s"]]), as.character, character(1))
    if (n > 0L && length(strs) >= n) {
      warning("64-bit integers beyond 2^53 are returned as doubles and lose ",
              "precision; the exact values are in info$s", call. = FALSE)
      return(as.numeric(strs[seq_len(n)]))
    }
  }

  vals <- .aslist(raw)
  filler <- if (logical_leaf) NA else NA_real_
  flat <- rep(filler, max(n, 0L))
  for (k in seq_len(min(n, length(vals)))) {
    v <- vals[[k]]
    if (!is.null(v)) {
      flat[k] <- if (logical_leaf) as.logical(v) else as.numeric(v)
    }
  }

  nf <- e[["nf"]]
  if (!is.null(nf) && !logical_leaf) {
    idxs <- as.integer(unlist(.aslist(nf[["i"]])))
    kinds <- vapply(.aslist(nf[["k"]]), as.character, character(1))
    for (q in seq_along(idxs)) {
      j <- idxs[q]
      if (is.na(j) || j < 1L || j > n) next
      flat[j] <- switch(kinds[q], "NaN" = NaN, "Inf" = Inf, "-Inf" = -Inf, flat[j])
    }
  }

  if (logical_leaf) as.logical(flat) else flat
}
